### ISSUE TITLE
[OP][Refactor] Merge native implementations of variance-related operators: BatchNorm, GroupNorm, LayerNorm, Moments

### ISSUE THREAD

**Issue by sxjscience** [apache/incubator-mxnet#15627]
There are lots of operators in MXNet that are related to calculating the forward + gradient w.r.t the variance, i.e., `var(X)`

- Moments
https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/moments.cc
- BatchNorm
https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/batch_norm.cc
- LayerNorm
https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/layer_norm.cc
- GroupNorm
https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/group_norm.cc

Some code bases can be merged to share the optimization tricks used in different places. So I propose to do some refactors.

We can also refer to the implementation of ATen https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Normalization.cpp

@pengzhao-intel @haojin2, what do you think?

**Comment by pengzhao-intel** [comment]
Good idea :)
Could you write up a simple description of what's the preferred optimization skill in these OPs?

 cc @TaoLv @ciyongch @zixuanweeei

**Comment by Jerryzcn** [comment]
@ptrendx @Kh4L Are you interested in optimizing GroupNorm kernel? It is used in the more recent RCNN models.

**Comment by ptrendx** [comment]
We are always interested in optimization opportunities :-). That said, we would need to prioritize it and I don't think we have resources for it currently. If somebody wants to try in the meantime I would be more than happy to talk and help shaping the GPU algorithm though.

**Comment by Jerryzcn** [comment]
> There are lots of operators in MXNet that are related to calculating the forward + gradient w.r.t the variance, i.e., `var(X, axis=...)`
> 
> * Moments
>   https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/moments.cc
> * BatchNorm
>   https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/batch_norm.cc
> * LayerNorm
>   https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/layer_norm.cc
> * GroupNorm
>   https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/group_norm.cc
> 
> Some code bases can be merged to share the optimization tricks used in different places. So I propose to do some refactors.
> 
> We can also refer to the implementation of ATen https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Normalization.cpp
> 
> Also, caffe2 https://github.com/pytorch/pytorch/blob/master/caffe2/operators/layer_norm_op.cc
> 
> @pengzhao-intel @haojin2, what do you think?

Can we add InstanceNorm as well thanks!
